<a href="https://colab.research.google.com/github/shamsheer-tech18/AppliedGenAI/blob/main/M3_Lab1_Prompting_Strategies.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<!-- Intro Section -->
<div style="background: linear-gradient(135deg, #001a70 0%, #0055d4 100%); color: white; padding: 30px; border-radius: 12px; text-align: center; box-shadow: 0 4px 12px rgba(0,0,0,0.1);">
    <h1 style="margin-bottom: 10px; font-size: 32px;">Introduction to Prompting Strategies</h1>
    <p style="font-size: 18px; margin: 0;">Instructor: <strong>Dr. Dehghani</strong></p>
</div>

<!-- Spacer -->
<div style="height: 30px;"></div>

<!-- Why It Matters Section -->
<div style="background: #ffffff; padding: 25px; border-radius: 10px; border-left: 6px solid #0055d4; box-shadow: 0 4px 8px rgba(0,0,0,0.05);">
    <h2 style="margin-top: 0; color: #001a70;">Why Prompting Strategies Matter</h2>
    <p style="font-size: 16px; line-height: 1.6;">
        Imagine you’re working with a junior engineer. You say:  
        <em>“Optimize the system.”</em><br>
        They’ll probably ask: <em>“Which system? Optimize for cost, speed, or energy? Any constraints?”</em> 🧐
    </p>
    <p style="font-size: 16px; line-height: 1.6;">
        Now try this instead:  
        <em>“Analyze the HVAC system and minimize energy consumption while keeping temperatures between 22-24°C. Provide a cost breakdown.”</em>  
    </p>
    <p style="font-size: 16px; line-height: 1.6;">
        That’s not just a prompt—it’s a <strong>clear strategy</strong> with defined objectives and boundaries.
        And that’s exactly what AI models need to perform at their best.
    </p>
</div>

<!-- Tip Section -->
<div style="background: #f5faff; padding: 20px; border-radius: 8px; border-left: 5px solid #0055d4; margin-top: 30px;">
    <h3 style="margin-top: 0; color: #0055d4;">💡 Pro Tip</h3>
    <p style="margin: 0; font-size: 16px; line-height: 1.6;">
        AI models appreciate well-structured instructions just like engineers appreciate complete design specs.
        Be specific, set clear goals, and watch the results improve!
    </p>
</div>

<!-- Upcoming Topics -->
<div style="margin-top: 40px; text-align: center;">
    <h3 style="color: #001a70;">What’s Ahead</h3>
    <ul style="list-style: none; padding: 0; font-size: 16px; line-height: 1.8;">
        <li>📚 Basic Prompting Types</li>
        <li>🧩 Advanced Strategies</li>
        <li>📊 Application-Specific Techniques</li>
    </ul>
    <p style="font-size: 16px; color: #333;">Let’s engineer some powerful AI conversations! 🛠️</p>
</div>


<!-- Section Header -->
<div style="background: linear-gradient(135deg, #001a70 0%, #0055d4 100%); color: white; padding: 25px; border-radius: 12px; text-align: center; box-shadow: 0 4px 12px rgba(0,0,0,0.1);">
    <h1 style="margin-bottom: 10px; font-size: 30px;">📚 Basic Prompting Types</h1>
</div>

<!-- Spacer -->
<div style="height: 25px;"></div>

<!-- Zero-Shot Prompting -->
<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4; margin-bottom: 20px;">
    <h3 style="margin-top: 0; color: #001a70;">1️⃣ Zero-Shot Prompting</h3>
    <p style="font-size: 16px; line-height: 1.6;">
        Provide only the task without any examples.  
        <strong>Use When:</strong> The task is simple and well-known by the model.  
        <em>Example:</em> “Translate 'Hello' to French.”
    </p>
</div>


In [6]:
# ==========================
# 📌 Set Up LLM and OpenAI API
# ==========================
# Import required libraries
from google.colab import userdata
import openai
import os

# Load the OpenAI API key securely from Colab secrets
api_key = userdata.get('OpenAI_API_Key')

# Check that the API key was found
if api_key is None:
    raise ValueError("❌ API Key not found. Please store your OpenAI API key using Colab secrets.")

# Set API key as environment variable for OpenAI
os.environ["OpenAI_API_Key"] = api_key

# Initialize OpenAI client
client = openai.OpenAI(api_key=api_key)

print("✅ OpenAI API Key successfully loaded and environment is ready!")

# ==========================
# 📌 Set LLM Model to GPT-3.5
# ==========================
# Define which LLM model to use
model_name = "gpt-3.5-turbo"

print(f"✅ LLM model set to: {model_name}")


✅ OpenAI API Key successfully loaded and environment is ready!
✅ LLM model set to: gpt-3.5-turbo


**OBSERVATION:**
In this task, provided the api_key for the model, and asked to check if the key is found with the provided info. Since it found the key, it has responded by printing the value as “OpenAI API Key successfully loaded and environment is ready!” and the LLM model was set to “gpt-3.5-turbo”

In [8]:
# ==========================
# 📌 Zero-Shot Test: Hidden Formula Sequence
# ==========================

hard_sequence_prompt_zero = (
    "The sequence is: 3, 12, 27, 48, 75, ___. What’s next?"
)

response_zero_hard = client.chat.completions.create(
    model="gpt-3.5-turbo",
    messages=[{"role": "user", "content": hard_sequence_prompt_zero}],
    temperature=0
)

print("🔹 LLM Response (Zero-Shot - Hard Sequence):\n")
print(response_zero_hard.choices[0].message.content.strip())


🔹 LLM Response (Zero-Shot - Hard Sequence):

The pattern in the sequence is adding consecutive odd numbers to the previous number. 

3 + 9 = 12
12 + 15 = 27
27 + 21 = 48
48 + 27 = 75

Therefore, the next number in the sequence would be 75 + 33 = 108. 

So, the next number in the sequence is 108.


**OBSERVATIONS:**
In this test, asked the model to find the hidden formula in the sequence numbers without providing any example to the model. The model responded by correctly identifying the pattern and predicted the next number as 108. This response was captured by setting the temperature value as “0”.


<!-- One-Shot Prompting -->
<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4; margin-bottom: 20px;">
    <h3 style="margin-top: 0; color: #001a70;">2️⃣ One-Shot Prompting</h3>
    <p style="font-size: 16px; line-height: 1.6;">
        Provide one clear example along with the instruction.  
        <strong>Use When:</strong> You want to guide the model’s behavior with a single example.  
        <em>Example:</em> “Translate 'Hello' to French: Bonjour. Now translate 'Goodbye'.”
    </p>
</div>


In [9]:
# ==========================
# 📌 Zero-Shot vs One-Shot Comparison: Alternating Pattern Sequence (Correct One-Shot)
# ==========================

model_name = "gpt-3.5-turbo"

# Zero-Shot Prompt (No Example)
zero_shot_prompt = (
    "The sequence is: 1, 4, 2, 9, 3, 16, 4, ___. What number should replace the blank?"
)

# One-Shot Prompt (One Example + New Question)
one_shot_prompt = (
    "Example:\n"
    "The sequence is: 1, 1, 2, 4, 3, 9, ___. What’s next?\n"
    "Answer: 4.\n\n"
    "Now solve this one:\n"
    "The sequence is: 1, 4, 2, 9, 3, 16, 4, ___. What number should replace the blank?"
)

# Run Zero-Shot
response_zero = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": zero_shot_prompt}],
    temperature=0
)

# Run One-Shot
response_one = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": one_shot_prompt}],
    temperature=0
)

# Display Results
print("🔹 Zero-Shot Response:\n" + "-"*40)
print(response_zero.choices[0].message.content.strip())

print("\n\n🔹 One-Shot Response:\n" + "-"*40)
print(response_one.choices[0].message.content.strip())


🔹 Zero-Shot Response:
----------------------------------------
The pattern is: 

1^2 = 1
2^2 = 4
3^2 = 9
4^2 = 16

Therefore, the next number in the sequence should be 5^2 = 25. 

So, the number that should replace the blank is 25.


🔹 One-Shot Response:
----------------------------------------
Answer: 25.


**OBSERVATIONS:**
Used “gpt-3.5-turbo” model
In this example, compared the Zero-shot and one-shot prompting, where in zero-shot asked the model to find the number in the sequence without giving any example, and in one-shot prompting asked model to find the number in the sequence by providing the examples to the model. In response, zero-shot provided the detailed explanation of why the next number is picked as 25, but in one-shot it gave the output value directly without explaining by referring the provided examples.



<!-- Few-Shot Prompting -->
<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4; margin-bottom: 20px;">
    <h3 style="margin-top: 0; color: #001a70;">3️⃣ Few-Shot Prompting</h3>
    <p style="font-size: 16px; line-height: 1.6;">
        Provide multiple examples to clearly demonstrate the pattern.  
        <strong>Use When:</strong> The task is complex or requires understanding a specific format.  
        <em>Example:</em>  
        - “Translate 'Hello' to French: Bonjour.”  
        - “Translate 'Goodbye' to French: Au revoir.”  
        - “Translate 'Thank you' to French: Merci.”  
        Now translate 'Good night'.
    </p>
</div>

<!-- Spacer -->
<div style="height: 30px;"></div>

<!-- Closing Tip -->
<div style="background: #f5faff; padding: 20px; border-radius: 8px; border-left: 5px solid #0055d4;">
    <h3 style="margin-top: 0; color: #0055d4;">💡 Quick Reminder</h3>
    <p style="margin: 0; font-size: 16px; line-height: 1.6;">
        The more complex the task, the more examples you should provide. But remember, too many examples can make prompts bulky and inefficient.
    </p>
</div>



In [11]:
# ==========================
# 📌 Few-Shot Prompting Example: Ultra-Hard Pattern (3 Hidden Rules)
# ==========================

model_name = "gpt-4-turbo"  # Best for complex reasoning

# Few-Shot Prompt with 2 Examples
few_shot_prompt = (
    "Example 1:\n"
    "The sequence is: 1, 1, 2, 4, 3, 9, ___. What’s next?\n"
    "Answer: 4.\n\n"
    "Example 2:\n"
    "The sequence is: 1, 1, 2, 4, 4, 9, 7, 16, ___. What’s next?\n"
    "Answer: 11.\n\n"
    "Now try this one:\n"
    "The sequence is: 1, 1, 2, 4, 4, 9, 7, 16, 11, ___, 16, 36. What number should replace the blank?"
)

# Run Few-Shot Prompt
response_few = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": few_shot_prompt}],
    temperature=0.5
)

# Display Result
print("🔹 Few-Shot Prompting (Two Examples Provided):")
print("-" * 40)
print(response_few.choices[0].message.content.strip())


🔹 Few-Shot Prompting (Two Examples Provided):
----------------------------------------
To solve this sequence, let's analyze the given numbers and see if we can identify a pattern or rule:

The sequence is: 1, 1, 2, 4, 4, 9, 7, 16, 11, ___, 16, 36.

Let's look at each number and see if there's a relationship between them:

1. The first few numbers don't provide a clear pattern, so let's look at the later numbers:
   - 4 (squared) = 16
   - 9 (squared) = 81, but we see 9 and then 7, which doesn't directly relate.
   - 4 (squared) = 16 (again)
   - 16 (squared) = 256, but we see 16 and then 36 (which is 6 squared).

2. We can observe that every few numbers, the sequence involves squares:
   - 4, 4, 9, 7, 16, 11, ___, 16, 36
   - Squares observed: 4, 9, 16, 36 (4^2, 3^2, 4^2, 6^2)

3. Let's see if there's a pattern in the numbers immediately before the squares:
   - Before 4 (which is 2^2), there is 2.
   - Before 9 (which is 3^2), there is 4.
   - Before 16 (which is 4^2), there is 7.
  

**OBSERVATIONS:**
Since we are solving a complex reasoning problem, used “gpt-4-turbo” model for solving it better.
In this example, provided 2 different examples of the reasoning problems and asked model to find a number in a sequence, and temperature was set to “0.5” to make the model respond accordingly. In response the model has defined the missing number as 16 and given the clear steps of identifying the missing number in the sequence. This is done as per the given examples to the model and to its reasoning ability.

## 🧠 Advanced Prompting Techniques  

Moving beyond basic prompting methods like zero-shot and few-shot, advanced strategies help enhance the reasoning and adaptability of large language models (LLMs). These techniques guide the model's thought process to handle complex tasks more effectively.

---

### 🔗 Chain-of-Thought (CoT) Prompting  

Chain-of-Thought prompting encourages models to **explain their intermediate reasoning steps**, leading to more transparent and accurate conclusions. By structuring prompts to include logical steps, CoT improves the model’s ability to solve complex reasoning tasks.

**Why is CoT Important?**  
- ✔️ Improves performance on multi-step reasoning tasks.  
- ✔️ Helps produce logically structured and coherent responses.  
- ✔️ Breaks down complex problems into manageable steps.

📖 **Reference:** [Chain-of-Thought Prompting Elicits Reasoning in Large Language Models](https://arxiv.org/abs/2201.11903)

---

*Next, explore practical examples of Chain-of-Thought prompting.*


In [12]:
# ==========================
# 📌 Chain-of-Thought Demonstration: Make 110 with Five 5's
# ==========================

model_name = "gpt-4-turbo"

# Zero-Shot Prompt (No Reasoning Encouraged)
zero_shot_prompt = (
    "Use exactly five 5’s and only four operations (+, -, *, /) and parentheses to make 110."
)

# Chain-of-Thought Prompt (Encourages Step-by-Step Reasoning)
cot_prompt = (
    "Let's solve this step by step.\n"
    "We need to use exactly five 5’s and only four operations (+, -, *, /) and parentheses to make 110.\n"
    "Step 1: Assume we cannot combine the 5's to form larger numbers (e.g., 55).\n"
    "Step 2: Try to combine them logically to reach 110.\n"
    "Now, provide the final equation and the answer."
)

# Run Zero-Shot
response_zero = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": zero_shot_prompt}],
    temperature=0
)

# Run Chain-of-Thought
response_cot = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": cot_prompt}],
    temperature=0
)

# Display Results
print("🔹 Zero-Shot Response (No Reasoning Encouraged):\n" + "-" * 50)
print(response_zero.choices[0].message.content.strip())

print("\n🔹 Chain-of-Thought Response (Reasoning Encouraged):\n" + "-" * 50)
print(response_cot.choices[0].message.content.strip())


🔹 Zero-Shot Response (No Reasoning Encouraged):
--------------------------------------------------
To achieve the number 110 using exactly five 5's and the operations (+, -, *, /) along with parentheses, you can arrange them as follows:

\[ 5 * (5 + 5) + 55 = 110 \]

Here's the breakdown:
1. \(5 + 5 = 10\)
2. \(5 * 10 = 50\)
3. \(50 + 55 = 110\)

This expression uses exactly five 5's and the operations as specified.

🔹 Chain-of-Thought Response (Reasoning Encouraged):
--------------------------------------------------
To solve this, we need to find a way to use exactly five 5's and the allowed operations to reach 110. Here's a step-by-step approach to find the solution:

Step 1: Start by considering the basic operations and how they can combine to reach a target close to 110 using five 5's.

Step 2: Experiment with different combinations of operations and groupings (parentheses) to manipulate the values effectively.

After trying different combinations, one possible solution is:
\[ (5 

# ✋ Hands-On Experiment: Observations  

📌 **Instructions:**  
- Run your experiments by changing the model type (e.g., `gpt-3.5-turbo`, `gpt-4-turbo`, `gpt-o3`), temperature, and prompt style.  
- You can **either attach a screenshot/image of your results** or **write a brief summary of your observations (max half a page)**.

---

- **Model Used:**  
  _[Enter the model name you tried, e.g., gpt-3.5-turbo, gpt-4-turbo, or gpt-o3]_

- **Temperature Setting:**  
  _[Enter the temperature you used, e.g., 0.0, 0.5, 0.7]_

- **Zero-Shot Result:**  
  _[Did Zero-Shot solve the problem correctly? Yes/No. Add a short explanation or attach an image.]_

- **Chain-of-Thought Result:**  
  _[Did Chain-of-Thought solve the problem better? Yes/No. Add a short explanation or attach an image.]_

- **Key Takeaways (Max Half Page or Screenshot):**  
  _[Summarize what you observed. Did a specific model perform better? How did temperature affect the results? What worked best? Attach image or write here.]_

---

✍️ *Try at least two models and different temperatures. Compare the results and reflect on how prompting strategies influence performance!*


In [14]:
# ===========================
# 🧪 Chain-of-Thought Demonstration: Make 24 with Four 3's
# ===========================

model_name = "gpt-4-turbo"

# Zero-Shot Prompt (No Reasoning Encouraged)
zero_shot_prompt = (
    "Use exactly four 3's and only four operations (+, -, *, /) and parentheses to make 24."
)

# Chain-of-Thought Prompt (Encourages Step-by-Step Reasoning)
cot_prompt = (
    "Let's solve this step by step.\n"
    "We need to use exactly four 3's and only four operations (+, -, *, /) and parentheses to make 24.\n"
    "Step 1: Assume we cannot combine the 3's to form larger numbers (e.g., 33).\n"
    "Step 2: Think about what operations on 3's can give us factors of 24.\n"
    "Step 3: Try to combine them logically to reach 24.\n"
    "Now, provide the final equation and the answer."
)

# Run Zero-Shot
response_zero = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": zero_shot_prompt}],
    temperature=0
)

# Run Chain-of-Thought
response_cot = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": cot_prompt}],
    temperature=0
)

# Display Results
print("◆ Zero-Shot Response (No Reasoning Encouraged):\n" + "-" * 50)
print(response_zero.choices[0].message.content.strip())

print("\n◆ Chain-of-Thought Response (Reasoning Encouraged):\n" + "-" * 50)
print(response_cot.choices[0].message.content.strip())

◆ Zero-Shot Response (No Reasoning Encouraged):
--------------------------------------------------
One way to use exactly four 3's and the operations (+, -, *, /) along with parentheses to make 24 is as follows:

\[ (3 \times 3 + 3) \times 3 = 24 \]

Here's the breakdown:
1. Multiply the first two 3's: \(3 \times 3 = 9\)
2. Add the third 3 to the result: \(9 + 3 = 12\)
3. Multiply the result by the fourth 3: \(12 \times 3 = 24\)

◆ Chain-of-Thought Response (Reasoning Encouraged):
--------------------------------------------------
To solve this, let's follow the steps you outlined and find a way to use four 3's and the operations (+, -, *, /) to make 24.

Step 1: We cannot combine the 3's to form larger numbers like 33.

Step 2: Consider operations that can give us factors or results that lead to 24. We know that:
- Multiplying 3 by 8 gives 24, but we need to see how to make 8 using three 3's.
- Adding 3 + 3 + 3 + 3 gives 12, which is half of 24.
- Dividing 3 by 3 gives 1, which might 

**OBSERVATIONS:**

**Model Used:** "gpt-4-turbo"

**Temperature:** "0"

**Zero-Shot Result:** Yes, this approach has solved the problem and given the answer directly but gave a brief breakdown of steps.

**CoT:** Yes, this too has solved the problem, also given step-by-step response. It explored multiple approached before concluding the answer. This model has shown how the model has evaluated different possibilities before concluding the answer.

**Key Take-Away:**
1. Both models produce correct answer, but solutions are different.
2. Zero-shot is faster and direct which can be suitable for simple problems.
3. CoT shows teh reasoning orocess, making it easy to understand the logic of the problem.
4. CoT approach will helps the model to explore teh multiple strategies before finalizing an answer.

```
# This is formatted as code
```



## 🔁 Self-Consistency Prompting

While Chain-of-Thought (CoT) improves reasoning by encouraging step-by-step thinking, it may still produce **inconsistent or incorrect** answers, especially in complex scenarios.  
**Self-Consistency Prompting** enhances CoT by asking the model to **generate multiple reasoning paths** and then select the most common or consistent final answer.

### Why is Self-Consistency Useful?

- ✅ Reduces random reasoning errors.
- ✅ Boosts reliability on ambiguous or multi-path problems.
- ✅ Often improves performance on mathematical, logical, and symbolic tasks.

📖 **Reference**: [Self-Consistency Improves Chain of Thought Reasoning in Language Models](https://arxiv.org/abs/2203.11171)

---

*Next, we’ll see how Self-Consistency works in action using a complex reasoning example.*


In [16]:
# ==========================
# 📌 Comparing Chain-of-Thought vs. Self-Consistency Prompting
# ==========================

model_name = "gpt-4-turbo"  # Using GPT-4 for better reasoning

# Define the problem prompt
problem_prompt = (
    "If a train travels at 60 miles per hour and leaves at 2 PM, and another train leaves "
    "the same station at 3 PM traveling at 90 miles per hour, when will the second train catch up to the first?"
)

# Chain-of-Thought Prompt (Standard)
cot_prompt = (
    "Let's solve this step by step.\n"
    + problem_prompt
)

# Self-Consistency Prompt: Ask the model to produce multiple reasoning paths
def run_self_consistency(prompt, num_attempts=5):
    answers = []
    for _ in range(num_attempts):
        response = client.chat.completions.create(
            model=model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.7  # Add randomness to explore different reasoning paths
        )
        answer = response.choices[0].message.content.strip()
        answers.append(answer)
    return answers

# Run Chain-of-Thought (Single Attempt)
response_cot = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": cot_prompt}],
    temperature=0
)
cot_answer = response_cot.choices[0].message.content.strip()

# Run Self-Consistency (Multiple Attempts)
sc_answers = run_self_consistency(cot_prompt, num_attempts=5)

# Simple Majority Vote to Find Most Consistent Answer
from collections import Counter
most_common_answer = Counter(sc_answers).most_common(1)[0]

# Display Results
print("🔹 Chain-of-Thought Response (Single Attempt):\n" + "-" * 50)
print(cot_answer)

print("\n🔹 Self-Consistency Responses (Multiple Attempts):\n" + "-" * 50)
for idx, ans in enumerate(sc_answers, 1):
    print(f"Attempt {idx}: {ans}")

print("\n🔹 Final Self-Consistency Selected Answer:\n" + "-" * 50)
print(f"Most Common Answer: {most_common_answer[0]}\nAppeared {most_common_answer[1]} times.")


🔹 Chain-of-Thought Response (Single Attempt):
--------------------------------------------------
To find out when the second train will catch up to the first, we can start by calculating how far ahead the first train is when the second train starts.

1. **Calculate the head start of the first train:**
   The first train leaves at 2 PM and travels at 60 miles per hour. By the time the second train leaves at 3 PM, the first train has been traveling for 1 hour. 
   Distance traveled by the first train in 1 hour = Speed × Time = 60 miles per hour × 1 hour = 60 miles.

2. **Set up the equation to find when the second train catches up:**
   Let \( t \) be the time in hours after 3 PM when the second train catches up to the first train. In this time, the first train travels an additional \( 60t \) miles (since it continues to travel at 60 mph), and the second train travels \( 90t \) miles (since it travels at 90 mph).

   Since the second train needs to cover the initial 60 miles gap plus the

<div style="background: linear-gradient(135deg, #001a70 0%, #0055d4 100%); color: white; padding: 25px; border-radius: 12px; text-align: center;">
    <h1 style="margin-bottom: 10px;">📚 Exploring More Advanced Prompting Strategies</h1>
</div>

<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4; margin-top: 20px;">
    <ul style="font-size: 16px; line-height: 1.8;">
        <li><strong>🧩 Tree-of-Thought (ToT) Prompting:</strong> Explores multiple reasoning paths like a decision tree, helping the model evaluate and compare various solutions before choosing the best one.</li>
        <li><strong>🤖 ReAct (Reasoning and Acting) Prompting:</strong> Combines reasoning steps with actions, including API calls or external tool usage. Ideal for interactive agents and dynamic decision-making tasks.</li>
        <li><strong>🔄 Reflexion Prompting:</strong> Encourages the model to critique its own responses and iteratively improve them, simulating self-correction and learning.</li>
    </ul>
</div>

<div style="margin-top: 40px; text-align: center;">
    <h2 style="color: #001a70;">✋ Hands-On Task: Compare Prompting Strategies</h2>
</div>

<div style="background: #f5faff; padding: 20px; border-radius: 8px; border-left: 5px solid #0055d4;">
    <p style="font-size: 16px;">
        📌 <strong>Task Instructions:</strong><br>
        - Experiment with <strong>Self-Consistency</strong>, <strong>Tree-of-Thought</strong>, and <strong>ReAct</strong> prompting methods.<br>
        - Try to solve the following problem using each method and compare the results.
    </p>
</div>

<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4; margin-top: 20px;">
    <h3>🧠 <strong>Challenge Problem:</strong></h3>
    <p style="font-size: 16px;">A farmer has chickens and rabbits in a cage. There are 35 heads and 94 legs. How many chickens and rabbits are there?</p>
</div>

<div style="margin-top: 40px;">
    <ul style="font-size: 16px; line-height: 1.8;">
        <li>Try different models (e.g., <code>gpt-3.5-turbo</code>, <code>gpt-4-turbo</code>, <code>gpt-o3</code>).</li>
        <li>Experiment with different temperatures (e.g., <code>0.0</code>, <code>0.5</code>, <code>0.7</code>).</li>
        <li>Use both direct prompts and advanced strategies like CoT, Self-Consistency, or ReAct.</li>
    </ul>
</div>

<div style="margin-top: 40px; text-align: center;">
    <h2 style="color: #001a70;">📖 Observations</h2>
</div>

<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4;">
    <ul style="font-size: 16px; line-height: 1.8;">
        <li><strong>Model and Strategy Used:</strong><br>_[Enter the model and prompting strategy you tried]_</li>
        <li><strong>Was the Correct Answer Found?</strong><br>_[Yes/No. Explain briefly or attach a screenshot]_</li>
        <li><strong>Key Takeaways (Max Half Page or Screenshot):</strong><br>_[Summarize how different strategies performed. What worked best? Why?]_</li>
    </ul>
</div>

<div style="margin-top: 20px; text-align: center;">
    ✍️ <em>Hint: Try breaking down the problem into equations or ask the model to explain its steps before giving the final answer. Notice which strategies lead to faster and more accurate results!</em>
</div>


In [18]:
# ==========================
# ✋ Hands-On Code: Try Different Prompting Strategies and Models
# ==========================

# 📝 Instructions:
# - Change 'model_name' to try different models (e.g., "gpt-3.5-turbo", "gpt-4-turbo", "gpt-o3").
# - Adjust 'temperature' to test how creativity affects reasoning.
# - Try Self-Consistency by sampling multiple outputs and comparing answers.
# - Optionally, explore Tree-of-Thought and ReAct patterns by modifying prompts.
# ✅ Your Experiment Starts Here 👇

# ===========================
# 🧠 Challenge Problem: Chickens and Rabbits
# Comparing Self-Consistency, Tree-of-Thought, and ReAct Prompting
# ===========================

import openai
from collections import Counter

# Initialize OpenAI client
client = openai.OpenAI(api_key=api_key)

# Define the problem
problem = "A farmer has chickens and rabbits in a cage. There are 35 heads and 94 legs. How many chickens and rabbits are there?"

# ===========================
# Strategy 1: Self-Consistency Prompting
# ===========================
model_name = "gpt-4-turbo"

def run_self_consistency(prompt, num_attempts=5):
    answers = []
    for _ in range(num_attempts):
        response = client.chat.completions.create(
            model=model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.7
        )
        answer = response.choices[0].message.content.strip()
        answers.append(answer)
    return answers

sc_prompt = (
    "Let's solve this step by step.\n"
    + problem
)

sc_answers = run_self_consistency(sc_prompt, num_attempts=5)
most_common_answer = Counter(sc_answers).most_common(1)[0]

print("◆ Self-Consistency Responses (5 Attempts):\n" + "-" * 50)
for idx, ans in enumerate(sc_answers, 1):
    print(f"Attempt {idx}: {ans}\n")

print("\n◆ Final Self-Consistency Answer:\n" + "-" * 50)
print(f"Most Common Answer appeared {most_common_answer[1]} times.")

# ===========================
# Strategy 2: Tree-of-Thought (ToT) Prompting
# ===========================
model_name = "gpt-4-turbo"

tot_prompt = (
    "Solve this problem by exploring multiple reasoning paths like a decision tree.\n\n"
    "Problem: " + problem + "\n\n"
    "Path 1: Use algebra (set up equations with variables).\n"
    "Path 2: Use guess and check method.\n"
    "Path 3: Use logical deduction starting from extreme cases.\n\n"
    "Evaluate each path, compare the solutions, and choose the best one.\n"
    "Provide the final answer with explanation."
)

response_tot = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": tot_prompt}],
    temperature=0
)

print("\n◆ Tree-of-Thought Response:\n" + "-" * 50)
print(response_tot.choices[0].message.content.strip())

# ===========================
# Strategy 3: ReAct (Reasoning and Acting) Prompting
# ===========================
model_name = "gpt-4-turbo"

react_prompt = (
    "Solve the following problem using the ReAct framework.\n"
    "For each step, provide:\n"
    "- Thought: What you're thinking\n"
    "- Action: What calculation or step you're taking\n"
    "- Observation: What result you got\n\n"
    "Problem: " + problem + "\n\n"
    "Continue until you reach the final answer."
)

response_react = client.chat.completions.create(
    model=model_name,
    messages=[{"role": "user", "content": react_prompt}],
    temperature=0
)

print("\n◆ ReAct Response:\n" + "-" * 50)
print(response_react.choices[0].message.content.strip())

# ===========================
# Bonus: Compare Different Models
# ===========================
models = ["gpt-3.5-turbo", "gpt-4-turbo"]
simple_cot_prompt = "Let's solve this step by step.\n" + problem

print("\n◆ Model Comparison (Chain-of-Thought):\n" + "-" * 50)
for model in models:
    response = client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": simple_cot_prompt}],
        temperature=0
    )
    print(f"\n{model}:\n{response.choices[0].message.content.strip()}\n")


◆ Self-Consistency Responses (5 Attempts):
--------------------------------------------------
Attempt 1: To solve this problem, let's use the number of heads and legs to set up a system of equations. Let \( c \) be the number of chickens and \( r \) be the number of rabbits.

1. Each animal has one head. So, the total number of heads gives us the equation:
   \[ c + r = 35 \]

2. Chickens have 2 legs each and rabbits have 4 legs each. The total number of legs gives us the equation:
   \[ 2c + 4r = 94 \]

Let's solve these equations step by step:

First, simplify the second equation by dividing every term by 2:
\[ c + 2r = 47 \]

Now you have two equations:
\[ c + r = 35 \]
\[ c + 2r = 47 \]

Subtract the first equation from the second to eliminate \( c \):
\[ (c + 2r) - (c + r) = 47 - 35 \]
\[ r = 12 \]

Now that we know \( r \) (the number of rabbits), substitute \( r = 12 \) back into one of the original equations to find \( c \) (the number of chickens). Using the first equation:
\[

Model and Strategy used for solving the problem statement:

**Models:** "gpt-3.5-turbo", "gpt-4-turbo"

**Strategies:** Self-consistency, Tree-ofoThoughts(ToT), and ReAct Prompting

Yes, the correct answer was found and it is 23 chickens and 12 rabbits. All the tree strtegies used were arrived at the correct solution.

**Take-Aways:**
1. Self-Consistency: It ran 5 attempts with temperature being set at 0.7, and majority voting confirmed teh answer, which concludes that this is reliable for validation purpose. This strategy can be best fit, when anyone wants to verify the accuract through multiple reasoning paths.
2. Tree-of-Thoughts (ToT): It has explored different paths like "algebra", "guess and check methods", "logical deduction from extreme cases", which later compared all the paths and chooses the best solution. This model is best fit for problems with multiple solution approaches.
3. ReAct: In this methos it has clear step-by-step process with explicit calculations to arrive on to a final answer. This can be best fit for interactive tasks which requires clear reasoning to find a solution.


Finally, it can be concluded that ToT and ReAct prompting will provide the most transparent reasoning, where as self-consistency is suitable for verifying the answers better. All these appraches works well for mathematical problems, but when compared, ToT offers a more comprehensive exploration of the solutions.

<div style="background: linear-gradient(135deg, #001a70 0%, #0055d4 100%); color: white; padding: 25px; border-radius: 12px; text-align: center;">
    <h1 style="margin-bottom: 10px;">📌 Conclusion</h1>
</div>

<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4; margin-top: 20px;">
    <p style="font-size: 16px; line-height: 1.8;">
        In this hands-on exploration, different advanced prompting strategies were tested to solve reasoning-based challenges.
        Through experimenting with <strong>Chain-of-Thought (CoT)</strong>, <strong>Self-Consistency</strong>, and other methods,
        the following key insights were observed:
    </p>
    <ul style="font-size: 16px; line-height: 1.8;">
        <li>Advanced prompting techniques significantly improve model performance, especially on complex, multi-step problems.</li>
        <li>Changing the <strong>model type</strong> and <strong>temperature</strong> can drastically affect reasoning quality and creativity.</li>
        <li>Some strategies, like <strong>Self-Consistency</strong>, help reduce random errors by exploring multiple reasoning paths.</li>
        <li>For ambiguous or challenging problems, combining strategies (e.g., CoT + Self-Consistency) often leads to the most reliable results.</li>
    </ul>
</div>

<div style="background: #f5faff; padding: 20px; border-radius: 8px; border-left: 5px solid #0055d4; margin-top: 20px;">
    <p style="font-size: 16px; font-style: italic;">
        📖 <em>Remember: Prompt engineering is both an art and a science. The more you experiment, the better you understand how to guide LLMs effectively!</em>
    </p>
</div>

<div style="margin-top: 40px; text-align: center;">
    <h3 style="color: #001a70;">✍️ Final Reflection</h3>
</div>

<div style="background: #ffffff; padding: 20px; border-radius: 10px; border-left: 6px solid #0055d4;">
    <p style="font-size: 16px;">
        _[Write 2-3 sentences summarizing what you personally learned about prompting strategies and how model selection or temperature influenced the results.]_
    </p>
</div>


**CONCLUSION:**

Through this hands-on lab course, I have learned that, the way we frame the questions in prompt to a LLM does matters a lot just as much as the question itself. For simple problem, a zero-shot prompts can works best, if the questions are basic, for complex problems techniques like the CoT, Self-consistency will helps the model to be able to reason more effectively.

Also, it is observed that, with the increase in temperature, the verity in responses was introduced. And also, choosing a right model will also plays a major role in handling the reasoing questions.

Overall, promoptig feels like, learning on how to communicate clearly with the AI. It is clear that, the more specific and structured we are with the instructions, then it is assured to have a better output with the AI models.